File: Week01/d5_matrix_addition.py
"""
Author: Penut Chen
Date: 2019-6-30
Description:
  Practice of matrix addition.
"""

import random

def main():
    a = Matrix(size=(3, 5), name='Matrix A')
    b = Matrix(size=(3, 5), name='Matrix B')

    print(a, end='\n\n')
    print(b, end='\n\n')
    print(a + b)

class Matrix:
    def __init__(self, matrix=None, size=None, name='matrix'):
        self.size = size
        self.row, self.col = size
        self.name = name
        if matrix is None:
            self.matrix = [[random.randint(10, 99) for i in range(self.col)] for j in range(self.row)]
        else:
            self.matrix = [[j for j in i] for i in matrix]

    def __str__(self):
        m = '\n'.join([' '.join(['%3d' % x for x in row]) for row in self.matrix])
        return '%s\n%s' % (self.name, m)

    def __add__(self, another):
        row, col = self.row, self.col
        m = [[self.matrix[i][j] + another.matrix[i][j] for j in range(col)] for i in range(row)]
        return Matrix(matrix=m, size=self.size, name='%s + %s' % (self.name, another.name))

if __name__ == "__main__":
    main()

File: Week01/d5_matrix_addition.c
/**
 * Author: Penut Chen
 * Date: 2019-6-30
 * Description: 
 *   Practice of matrix addition.
 */

#include <stdio.h>
#include <stdlib.h>

int** generate_matrix(int, int);
void display_matrix(int**, int, int);
int** matrix_addition(int**, int**, int, int);

int main() {
    int row = 3, col = 5;
    int **a = generate_matrix(row, col);
    int **b = generate_matrix(row, col);

    printf("Matrix A\n");
    display_matrix(a, row, col);

    printf("\nMatrix B\n");
    display_matrix(b, row, col);

    int **c = matrix_addition(a, b, row, col);
    printf("\nA + B = \n");
    display_matrix(c, row, col);

    free(a);
    free(b);
    free(c);
    
    return 0;
}

int** generate_matrix(int row, int col) {
    int **matrix = (int**)malloc(sizeof(int*) * row);
    for (int i = 0; i < row; i++) {
        matrix[i] = (int*)malloc(sizeof(int) * col);
        for (int j = 0; j < col; j++)
            matrix[i][j] = rand() % 90 + 10;
    }
    return matrix;
}

void display_matrix(int **matrix, int row, int col) {
    for (int i = 0; i < row; i++) {
        for (int j = 0; j < col; j++)
            printf("%3d ", matrix[i][j]);
        printf("\n");
    }
}

int** matrix_addition(int **a, int **b, int row, int col) {
    int **result = (int**)malloc(sizeof(int*) * row);

    for (int i = 0; i < row; i++) {
        result[i] = (int*)malloc(sizeof(int) * col);
        for (int j = 0; j < col; j++)
            result[i][j] = a[i][j] + b[i][j];
    }

    return result;
}